### ISSUE TITLE
Redis Cache error

### ISSUE THREAD

**Issue by MarekKarmelski** [magento/magento2#9875]
### Preconditions

1. Magento 2.1.4 / 2.1.7
2. Default store

### Steps to reproduce

1. Use redis server to cache

### Expected result

1. Redis cache shoud work good.

### Actual result

1. Page not working
2. [2017-06-08 10:22:47] main.CRITICAL: exception 'Exception' with message 'Warning: Redis::pipeline(): Already in pipeline mode in /vendor/colinmollenhour/credis/Client.php on line 1037'

**Comment by orlangur** [comment]
More scenarios here: https://github.com/phpredis/phpredis/issues/1180

**Comment by MarekKarmelski** [comment]
So where is the problem in Redis serwer or Magento configuration?

**Comment by orlangur** [comment]
I didn't analyze the code at all, it looks to me that bug is in Magento code or code of `colinmollenhour/credis/Client.php`, maybe it occurs only in some specific configuration.

**Comment by MarekKarmelski** [comment]
Yeap, on my local server everything work perfect, on test/production server no. Locally I have Ubuntu, test/production CentOS.

**Comment by davidalger** [comment]
@MarekKarmelski Do you have the php-redis extension installed? We deploy to CentOS / RHEL using redis with no trouble here.

**Comment by MarekKarmelski** [comment]
@davidalger yeap, I have. I found what was the problem, our server administrator set autoupdating, and dwo days ago wird extension, was installed to redis server. When he removed and reinstalled redis one more time, the problem has disappeared. I thin that we can close this ticket.

**Comment by orlangur** [comment]
Cool, thanks for the quick investigation on your side :+1:

**Comment by mattias** [comment]
Getting the same issue on many centos7 servers right now.
On one redis had been working for months and after I refreshed the caches it stopped working with this particular issue.
On another, I just tried to do a "magento setup:backup --db" and I got the error only there but the site was working, haven't tried clearing all the caches on that one yet.

All servers do have autoupdating on and I have tried reinstalling redis as well as manually installing redis 4 but the problem remains.

The only solution I have so far is to remove redis from the magento 2 configuration. Effectively running without redis.

**Comment by mattias** [comment]
I found the issue.
2017-07-18 04:22 there was an automatic update for php-redis.
Package php-pecl-redis.x86_64 0:3.1.2-1.el7.remi.7.0
upgraded to 
Package php-pecl-redis.x86_64 0:3.1.3-1.el7.remi.7.0

Maybe this only happens to systems running the remi repository on centos7?

I downgraded php-pecl-redis and now it works again.

**Comment by chaohe** [comment]
We ran into the same problem after an automatic update to phpredis 3.1.3. @yatsukhnenko mentions over there that it was a purposeful change that caused side effects, may get removed in next phpredis. If not, may need an update in https://github.com/colinmollenhour/credis/blob/master/Client.php

**Comment by doubouil** [comment]
The error was solved for me with the lastest Credis Client @chaohe linked, thanks !

**Comment by anjanbarik** [comment]
I faced the same issue.  Php Redis Upgraded automatically from 3.1.2 to 3.1.3 and as the PHP-FPM was not restarted, so in phpinfo it was showing 3.1.2 version. Then I restarted PHP-FPM then it shows 3.1.3 version, then I downgraded it to 3.1.2 and issue resolved.

**Comment by stansm** [comment]
Faced the same issue. For a hotfix use [this commit](https://github.com/colinmollenhour/credis/commit/74b2b703da5c58dc07fb97e8954bc63280b469bf#diff-d506ea9bbbc4f3d805844520d3ec0efe)
or upgrade credis if possible

**Comment by patsky1980** [comment]
This is Credis bug: 2 times calling the pipeline method. Please, patche the file vendor/colinmollenhour/credis/Client.php by the code:

```php
            // Proxy pipeline mode to the phpredis library
            if($name == 'pipeline' || $name == 'multi') {
                if($this->isMulti) {
                    return $this;
                } else {
                    $this->isMulti = TRUE;
                    $this->redisMulti = call_user_func_array(array($this->redis, $name), $args);
                    //CUSTOM CODE!!!
                    //Credis bug fix: 2 times pipeline() calls below.
                    return $this;
                    //custom code ends
                }
            }
```

in lines 1013-1025

**Comment by orlangur** [comment]
@patsky1980 patching vendors is (almost) never a good idea, are there any issues if upgrade to `1.7` `colinmollenhour/credis` version (the one containing commit with fix)?

**Comment by dbrekelmans** [comment]
This should be fixed in 1.7: https://github.com/colinmollenhour/credis/commit/74b2b703da5c58dc07fb97e8954bc63280b469bf

**Comment by dbrekelmans** [comment]
This is fixed in 1.7: https://github.com/colinmollenhour/credis/commit/74b2b703da5c58dc07fb97e8954bc63280b469bf

**Comment by EliasKotlyar** [comment]
We can still reproduce this in 2.2

**Comment by orlangur** [comment]
@EliasKotlyar fix is already in `2.2-develop` but not released yet.
